datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -534,6 +574,655 @@ impl LogicalPlan {
             _ => {}
         }
     }
+
+    /// recursively to replace the params (e.g $1 $2, ...) wit corresponding values provided in the prams_values
+    pub fn replace_params_with_values(
+        &self,
+        param_values: &Vec<ScalarValue>,
+    ) -> Result<LogicalPlan, DataFusionError> {
```

Comment:
This is actually create a new logical plan. I hope sometime in the future someone will optimize this to only create expressions for placeholders/params and keep the rest the same. Not sure if that is possible/easy though

Reply:
I think you can use `datafusion_expr::utils::from_plan`  to avoid most of this code and make this PR quite a bit smaller

https://docs.rs/datafusion-expr/15.0.0/datafusion_expr/utils/fn.from_plan.html